[Tree] Add alternate rows for readability.

WhalesState opened the pull request:
Implements: https://github.com/godotengine/godot/pull/111268#issuecomment-3368760141

<img width="1919" height="1070" alt="image" src="https://github.com/user-attachments/assets/3de23536-b296-4c23-adf5-ecd775bbb242" />

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/9376

KoBeWi commented:
#89890 🙃

Mickeon commented:
Why is `draw_color_lines` an integer and doesn't allow a color to be chosen like https://github.com/godotengine/godot/pull/89890? It's somewhat true that one can get an acceptable value automatically, but the choice is important for most users and developers alike.
Regardless, my personal consensus is still the same as https://github.com/godotengine/godot/pull/89890#issuecomment-2020262693

ettiSurreal commented:
How would this look on the [new editor theme](https://github.com/godotengine/godot/pull/111118)? I think it could address the complaint of the filesystem and scene docks not having a background.

AdriaandeJongh reviewed (approved):
This will greatly improve the readability of trees / tables in the editor.

Didn't look at the code.

AdriaandeJongh reviewed (approved):
I tried the PR (didn't look at the code) and went over some of the trees the alternating row colors were applied to, and it is indeed very nice on those tables. It really does help with readability of wide tree items.

I also agree that this should be disabled by default.

Are these color guide colors exposed though style boxes when users add their own Tree nodes and style them? I couldn't find this out quickly in my test project.

IMO we should be careful not to add it to too many trees; especially ones with only one column and that aren't very wide. I think trees like the SceneTree and FileSystem do not need the alternating row colors because they are not wide by default and don't really behave as tables. @passivestar will have an opinion on this that may be more specific than mine, but IMO from what I checked the ones you already applied this to seem fine.

KoBeWi commented:
> Right now this can be enabled with a single line of code
tree->add_theme_constant_override("draw_color_guides", 1)
If we switch to using styleboxes, managing this becomes more complicated. Most styleboxes are disabled by default using EmptyStyleBox, so to keep the current default behavior we would need the default stylebox to remain empty.

Right now the guides can be customized with a single color. This could be changed to a StyleBox, separately from `draw_color_guides`, which would still toggle it.

I also wonder if `draw_color_guides` should even be a theme property if it's not something you'd want to have consistently enabled/disabled for every tree. Could be a regular property.

WhalesState committed:
[Tree] Add color guides.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

The feature is off by default, so that projects upgrading don't have their theming changed.

Needs a rebase.

### Modern

Dark | Light
-|-
<img width="1350" height="260" alt="Image" src="https://github.com/user-attachments/assets/9267c6ea-9b9c-451c-94d2-d4de67245bed" /> | <img width="1350" height="260" alt="Image" src="https://github.com/user-attachments/assets/60123d2a-a474-4b31-8fd8-ac7e750c0800" />

### Classic

Dark | Light
-|-
<img width="1350" height="296" alt="Image" src="https://github.com/user-attachments/assets/f0e123ab-4f25-47b4-973c-0a621ae0bb79" /> | <img width="1350" height="296" alt="Image" src="https://github.com/user-attachments/assets/b2ef96fc-5451-4c49-95d9-cfe9edbe1e79" />

WhalesState commented:
Rebased. Thanks for testing the new changes.

KoBeWi commented:
The new GDExtension tab in Project Settings should also have this enabled.

KoBeWi reviewed (approved):
The implementation looks fine, but I wonder about the property name. Maybe something like `use_alternate_rows` or `use_alternating_row_style` (or similar) would be more appropriate.

WhalesState commented:
`alternate_row_style_visible` ?

Edit:
`alternate_rows` may also work if anyone prefers short names.

Cautami commented:
Is it correct that this isn't applied to Trees like the FileSystem and the SceneTree view by default? 
I've tried this PR and while I do see the option for enabling alternating lines for Tress I make in the engine, I can't find anyway to actually enable this for these editor views in particular, which is what I most was looking forward to. 

I understand that the usability of this mostly comes from when it's a multi-column type of Tree, but I do think there's a certain level of added "niceness" in having stripes for even simple views. Blender for instance isn't much bigger than how I have mine setup in Godot, and while the effect is much more subtle than the proposed look here, not to be greedy or anything, but I would be appreciative of a toggle that allowed for it, and if not that's okay too. 

<img width="916" height="365" alt="side-by-side" src="https://github.com/user-attachments/assets/db06554a-ed8c-479d-93cd-b8aa6f2f53af" />

Unless of course there already is, in which case I'm surely blind because I cannot seem to find anything alluding to it in either the code or the editor settings :sweat: 

As for the naming of the property... while I don't believe I'm any source of authority, I'm partial towards `alternate_row_style_visible`, I don't think prefixing with `use_` makes sense, given the context of the other property just above (`column_titles_visible`), which is functionally similar to the alternate row property, and doesn't have `use_`. Would be inconsistent, I think.

KoBeWi commented:
My problem with `alternate_rows_visible` is that it's ambiguous. It may be interpreted as a toggle for the whole rows, not the style. Something that better describes it visually would be better IMO, like `stripes_visible` even.

It's just a suggestion though, going with the current name is ok too 🤷‍♂️

AdriaandeJongh commented:
`alternate_item_background_style`
`alternate_item_backgrounds`

WhalesState commented:
I think `striped_rows` is more descriptive than `alternate_rows`.

`striped_rows_visible` for the property.
`striped_rows_background` for the theme stylebox property.

WhalesState commented:
Renamed the property to `row_stripes_visible` to match `column_titles_visible` name style.

Enabled it for the new `ProjectSettingsGDExtension` tab, and added an option to enable it for the scene tree dock.

WhalesState commented:
> Is it correct that this isn't applied to Trees like the `FileSystem` and the `SceneTree` view by default? I've tried this PR and while I do see the option for enabling alternating lines for Tress I make in the engine, I can't find anyway to actually enable this for these editor views in particular, which is what I most was looking forward to.

Hello, I have added an option in EditorSettings `docks/scene_tree/show_row_stripes` to allow enabling this new feature for scene tree dock.

For `EditorFileSystem`, it can be added later, since it doesn't go well with folder background colors.

Cautami commented:
> Hello, I have added an option in EditorSettings `docks/scene_tree/show_row_stripes` to allow enabling this new feature for scene tree dock.
> 
> For `EditorFileSystem`, it can be added later, since it doesn't go well with folder background colors.

I've been using this PR for a couple weeks now and this has very much helped with readability with the scene tree, thank you. 
However, I have ran into an issue regarding the color of it, as with certain background colors the alternate row becomes completely invisible, and with the background color I normally use (#1e1e2e) it is quite hard to see.
I'm not entirely sure how this should be handled as I'm not sure if a new color setting in the editor theme settings is something desirable.

With a fully black background, or otherwise referred to as the "OLED" theme, the stripe has no actual appearance.
<img width="431" height="99" alt="image" src="https://github.com/user-attachments/assets/2f5104fc-b86e-4559-afef-97ebe6d0acdf" />
<img width="2118" height="290" alt="image" src="https://github.com/user-attachments/assets/ea73e6ee-caad-4e0d-b7a0-9f950021bd53" />

Additionally, this may seem a bit nit picky, but for both the editor and the actual theme usage of this, would it be possible to have an option or way of extending the alternate row through out the entirety of the panel?
 
As seen in Blender, I find this to be a more visually pleasing approach as it gives the look that each alternating row is a "slot" so to speak for something to go inside of it, I'm not sure how many people actually care about this sort of thing so I understand if it's not something you want to do. (And perhaps it could be something I can look into otherwise)
<img width="443" height="300" alt="image" src="https://github.com/user-attachments/assets/3dfc9694-c38f-41ff-b102-4fabb7a1fde8" />

I once again thank you for your work, and I hope to see this new style option become used elsewhere in the editor in the future, especially with the FileSystem dock. 

---

I went ahead and modified this to allow an alternate_row_color within interface/theme
It does feel slightly overkill having a new color option here, but given it's mostly a readability/accessibility feature perhaps it's okay? 
As an aside, I think the alternate rows, specifically within the tree view, are visually better if the rounded corners are removed.
Here's what a catppuccin-like theme would look like before and after colored alternate rows and no rounded corners.
<img width="423" height="139" alt="image" src="https://github.com/user-attachments/assets/b3e8b371-fded-48c1-9c20-af8bd2c34664" />

<img width="423" height="142" alt="image" src="https://github.com/user-attachments/assets/b2ad1ffe-f2b5-4cee-a93b-8d7be5304c85" />

A minor difference, but when the tree view is at the corner of a screen where the color is slightly shifted, it actually does help quite a bit.
If this is something desirable I can add presets for each theme (currently all just empty) and give you a .patch

Repiteo commented:
@WhalesState Is the above comment something you're interested in integrating? The PR will need a rebase anyway, so it'd probably be worth a look

WhalesState commented:
Yes, the current dark theme alpha value is `0.025` which only works with darkest colors, I'll need to adjust it somehow based on theme `base_color` or we allow the users to adjust it with a slider. I can rebase it tomorrow then we test it with different theme colors to find the best alpha value for dark themes or we allow adjusting it from editor theme settings.

WhalesState commented:
Added `strip_alpha` in Editor theme settings to control the opacity of the row strips.

Cautami commented:
Tested it and it definitely resolves that issue, thanks!

Though with Sticky Trees merged this seems to have created a more notable issue where the alternating lines seem to dance back and forth on who gets what

https://github.com/user-attachments/assets/ddb085c2-55db-4458-8a49-2d67ba3df33d

WhalesState commented:
Fixed, Thanks you ^

https://github.com/user-attachments/assets/48cf56b7-59fb-4764-89f7-7947a44ade91

The sticky items now renders with the same strip it had.

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -1286,6 +1289,9 @@
 			The editor theme spacing preset to use. See also [member interface/theme/base_spacing] and [member interface/theme/additional_spacing].
 		</member>
+		<member name="interface/theme/strip_alpha" type="float" setter="" getter="">
+			The opacity to use when drawing row strips and row guides in the editor's [Tree]-based GUIs (such as the Scene tree dock).
```
- I am well-aware that it is often referred as the Scene tree dock, but it truly is named "Scene dock".
- I am not sure what "row guides" fully entails here. For most intents and purposes it seems to be the same thing as "row strips". But please do correct me.


```suggestion
			The opacity to use when drawing striped rows for any interface based on [Tree] in the editor, such as the Scene dock. If set to [code]0.0[/code], no stripes will be drawn. See also [member docks/scene_tree/show_row_stripes].
```

Tangentially related: the name of this setting is a bit misleading. "Strip" sounds like a command, as if what the setting does is cut off anything whose alpha passes a certain threshold, for some reason.

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -326,6 +326,9 @@
 			If [code]true[/code], the scene tree dock will only show nodes that match the filter, without showing parents that don't. This settings can also be changed in the Scene dock's top menu.
 		</member>
+		<member name="docks/scene_tree/show_row_stripes" type="bool" setter="" getter="">
+			If [code]true[/code], the scene tree dock will draw striped rows to improve readability.
```
```suggestion
			If [code]true[/code], the Scene dock will draw striped rows to improve readability. See also [member interface/theme/strip_alpha].
```

Mickeon commented on scene/gui/tree.cpp:
```diff
@@ -7518,6 +7549,7 @@ void Tree::_bind_methods() {
 
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "columns"), "set_columns", "get_columns");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "column_titles_visible"), "set_column_titles_visible", "are_column_titles_visible");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "row_stripes_visible"), "set_row_stripes_visible", "are_row_stripes_visible");
```
Nnnnno. I am a bit concerned about this. I'm pretty sure this should be a theme item, but @godotengine/gui-nodes would have to back me up on this. My reasoning is:
- This is a rather notable aesthetic change that _I would imagine_ most users would rather have propagate across a whole theme.
- The `row_stripes` stylebox, which depends on this property, is already a theme item. So one "variable" propagates, but the other doesn't and still needs to be enabled on a case-by-case basis.
	- For more granularity it's still possible to override theme items for each Tree.

Note that we are no strangers to using theme constants as makeshift boolean properties (`0` for `false`, anything else is `true`).
See `draw_relationship_lines` and `draw_guides` on **Tree** itself. One could argue this property should be a theme item named `draw_stripes` or similar, instead.
